core/src/subgraph/instance_manager.rs

```diff
@@ -317,3 +235,528 @@ impl EventConsumer<SubgraphAssignmentProviderEvent> for SubgraphInstanceManager
+            persist_dynamic_data_sources(
+                logger3,
+                subgraph_state,
+                block_state,
+                data_sources,
+                runtime_hosts,
+                block_ptr_for_new_data_sources,
+            )
+            .map(move |(subgraph_state, block_state, data_sources_created)| {
+                // If new data sources have been added, indicate that the subgraph
+                // needs to be restarted after this block
+                if data_sources_created {
+                    needs_restart.swap(true, Ordering::SeqCst);
+                }
+
+                (subgraph_state, block_state)
+            })
+            .from_err()
+        },
+    )
+    // Apply entity operations and advance the stream
+    .and_then(move |(subgraph_state, block_state)| {
+        // Avoid writing to store if block stream has been canceled
+        if block_stream_cancel_handle.is_canceled() {
+            return Err(CancelableError::Cancel);
+        }
+
+        info!(
+            logger4,
+            "Applying {} entity operation(s)",
+            block_state.entity_operations.len()
+        );
+
+        // Transact entity operations into the store and update the
+        // subgraph's block stream pointer
+        subgraph_state
+            .store
+            .transact_block_operations(
+                subgraph_state.deployment_id.clone(),
+                block_ptr_now,
+                block_ptr_after,
+                block_state.entity_operations,
+            )
+            .map(|_| subgraph_state)
+            .map_err(|e| {
+                format_err!("Error while processing block stream for a subgraph: {}", e).into()
+            })
+    })
+}
+
+fn process_logs<T>(
+    logger: Logger,
+    instance: Arc<RwLock<SubgraphInstance<T>>>,
+    block_state: BlockState,
+    block: Arc<EthereumBlock>,
+    logs: Vec<Log>,
+) -> impl Future<Item = BlockState, Error = CancelableError<Error>>
+where
+    T: RuntimeHostBuilder,
+{
+    stream::iter_ok::<_, CancelableError<Error>>(logs)
+        // Process events from the block stream
+        .fold(block_state, move |block_state, log| {
+            let logger = logger.clone();
+            let instance = instance.clone();
+            let block = block.clone();
+
+            let transaction = block
+                .transaction_for_log(&log)
+                .map(Arc::new)
+                .ok_or_else(|| format_err!("Found no transaction for event"));
+
+            future::result(transaction).and_then(move |transaction| {
+                instance
+                    .read()
+                    .unwrap()
+                    .process_log(&logger, block, transaction, log, block_state)
+                    .map(|block_state| block_state)
+                    .map_err(|e| format_err!("Failed to process event: {}", e))
+            })
+        })
+}
+
+fn process_logs_in_runtime_hosts<R>(
```

Comment:
Good idea. I didn't think it would be easy but it was. First time I've defined a static method on a trait, I think:
```
SubgraphInstance::<T>::process_log_in_runtime_hosts(...)
```
Fun!

Here's the commit that I added: https://github.com/graphprotocol/graph-node/commit/3808bb5082b56982142f4a0761cda51fad2b0506

Reply:
Nice!